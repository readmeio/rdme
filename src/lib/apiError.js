module.exports = class extends Error {
  constructor(res) {
    let err;

    // Special handling to for fetch `res` arguments where `res.error` will contain our API error response.
    if (typeof res === 'object') {
      if (typeof res?.error === 'object') {
        err = res.error;
      } else {
        err = res;
      }
    } else {
      err = res;
    }

    super(err);

    if (typeof err === 'object') {
      this.code = err.error;

      // If we returned help info in the API, show it otherwise don't render out multiple empty lines as we sometimes
      // throw `Error('non-api custom error message')` instances and catch them with this class.
      if (err?.help) {
        this.message = [err.message, '', err.help].join('\n');
      } else {
        this.message = err.message;
      }

      this.name = 'APIError';
    } else {
      this.code = err;
      this.message = err;
    }
  }
};
