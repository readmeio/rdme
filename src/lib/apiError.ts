interface APIErrorResponse {
  docs?: string;
  error: string;
  help?: string;
  message: string;
  poem?: string[];
  suggestion?: string;
}

export default class APIError extends Error {
  code: string;

  constructor(res: string | APIErrorResponse | { error: APIErrorResponse }) {
    let err: string | APIErrorResponse;

    // Special handling to for fetch `res` arguments where `res.error` will contain our API error
    // response.
    if (typeof res === 'object') {
      if (typeof res?.error === 'object') {
        err = res.error;
      } else {
        err = res as APIErrorResponse;
      }
    } else {
      err = res;
    }

    super(err as unknown as string);

    this.name = 'APIError';

    if (typeof err === 'object') {
      this.code = err.error;

      // If we returned help info in the API, show it otherwise don't render out multiple empty
      // lines as we sometimes throw `Error('non-api custom error message')` instances and catch
      // them with this class.
      if (err?.help) {
        this.message = [err.message, '', err.help].join('\n');
      } else {
        this.message = err.message;
      }

      this.name = 'APIError';
    } else {
      this.code = err;
      this.message = err;
    }
  }
}
