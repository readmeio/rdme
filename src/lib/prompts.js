const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
