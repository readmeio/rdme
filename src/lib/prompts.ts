import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
