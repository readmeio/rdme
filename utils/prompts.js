exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
