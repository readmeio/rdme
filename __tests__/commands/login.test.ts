import nock from 'nock';
import prompts from 'prompts';
import { describe, beforeAll, afterAll, afterEach, it, expect } from 'vitest';

import Command from '../../src/commands/login.js';
import { APIv1Error } from '../../src/lib/apiError.js';
import configStore from '../../src/lib/configstore.js';
import { getAPIV1Mock } from '../helpers/get-api-mock.js';
import { runCommandAndReturnResult } from '../helpers/oclif.js';

const apiKey = 'abcdefg';
const email = 'user@example.com';
const password = 'password';
const project = 'subdomain';
const token = '123456';

describe('rdme login', () => {
  let run: (args?: string[]) => Promise<string>;

  beforeAll(() => {
    nock.disableNetConnect();
    run = runCommandAndReturnResult(Command);
  });

  afterEach(() => configStore.clear());

  afterAll(() => nock.cleanAll());

  it('should error if no project provided', () => {
    prompts.inject([email, password]);
    return expect(run()).rejects.toStrictEqual(new Error('No project subdomain provided. Please use `--project`.'));
  });

  it('should error if email is invalid', () => {
    prompts.inject(['this-is-not-an-email', password, project]);
    return expect(run()).rejects.toStrictEqual(new Error('You must provide a valid email address.'));
  });

  it('should post to /login on the API', async () => {
    prompts.inject([email, password, project]);

    const mock = getAPIV1Mock().post('/api/v1/login', { email, password, project }).reply(200, { apiKey });

    await expect(run()).resolves.toBe('Successfully logged in as user@example.com to the subdomain project.');

    mock.done();

    expect(configStore.get('apiKey')).toBe(apiKey);
    expect(configStore.get('email')).toBe(email);
    expect(configStore.get('project')).toBe(project);
  });

  it('should post to /login on the API if passing in project via opt', async () => {
    prompts.inject([email, password]);

    const mock = getAPIV1Mock().post('/api/v1/login', { email, password, project }).reply(200, { apiKey });

    await expect(run(['--project', project])).resolves.toBe(
      'Successfully logged in as user@example.com to the subdomain project.',
    );

    mock.done();

    expect(configStore.get('apiKey')).toBe(apiKey);
    expect(configStore.get('email')).toBe(email);
    expect(configStore.get('project')).toBe(project);
  });

  it('should bypass prompts and post to /login on the API if passing in every opt', async () => {
    const mock = getAPIV1Mock().post('/api/v1/login', { email, password, project, token }).reply(200, { apiKey });

    await expect(run(['--email', email, '--password', password, '--project', project, '--otp', token])).resolves.toBe(
      'Successfully logged in as user@example.com to the subdomain project.',
    );

    mock.done();

    expect(configStore.get('apiKey')).toBe(apiKey);
    expect(configStore.get('email')).toBe(email);
    expect(configStore.get('project')).toBe(project);
  });

  it('should bypass prompts and post to /login on the API if passing in every opt (no 2FA)', async () => {
    const mock = getAPIV1Mock().post('/api/v1/login', { email, password, project }).reply(200, { apiKey });

    await expect(run(['--email', email, '--password', password, '--project', project])).resolves.toBe(
      'Successfully logged in as user@example.com to the subdomain project.',
    );

    mock.done();

    expect(configStore.get('apiKey')).toBe(apiKey);
    expect(configStore.get('email')).toBe(email);
    expect(configStore.get('project')).toBe(project);
  });

  it('should error if invalid credentials are given', async () => {
    prompts.inject([email, password, project]);
    const errorResponse = {
      error: 'LOGIN_INVALID',
      message: 'Either your email address or password is incorrect',
      suggestion: 'You can reset your password at https://dash.readme.com/forgot',
      help: 'If you need help, email support@readme.io and mention log "fake-metrics-uuid".',
    };

    const mock = getAPIV1Mock().post('/api/v1/login', { email, password, project }).reply(401, errorResponse);

    await expect(run()).rejects.toStrictEqual(new APIv1Error(errorResponse));
    mock.done();
  });

  it('should make additional prompt for token if login requires 2FA', async () => {
    prompts.inject([email, password, project, token]);
    const errorResponse = {
      error: 'LOGIN_TWOFACTOR',
      message: 'You must provide a two-factor code',
      suggestion: 'You can do it via the API using `token`, or via the CLI using `rdme login --2fa`',
      help: 'If you need help, email support@readme.io and mention log "fake-metrics-uuid".',
    };

    const mock = getAPIV1Mock()
      .post('/api/v1/login', { email, password, project })
      .reply(401, errorResponse)
      .post('/api/v1/login', { email, password, project, token })
      .reply(200, { apiKey });

    await expect(run()).resolves.toBe('Successfully logged in as user@example.com to the subdomain project.');

    mock.done();

    expect(configStore.get('apiKey')).toBe(apiKey);
    expect(configStore.get('email')).toBe(email);
    expect(configStore.get('project')).toBe(project);
  });

  it('should error if trying to access a project that is not yours', async () => {
    const projectThatIsNotYours = 'unauthorized-project';
    prompts.inject([email, password, projectThatIsNotYours]);
    const errorResponse = {
      error: 'PROJECT_NOTFOUND',
      message: `The project (${projectThatIsNotYours}) can't be found.`,
      suggestion: `The project is referred to as your \`subdomain\` in the dashboard when you're looking for it. If you're sure it exists, maybe you don't have access to it? You can check if it exists here: https://${projectThatIsNotYours}.readme.io.`,
      help: 'If you need help, email support@readme.io',
    };

    const mock = getAPIV1Mock()
      .post('/api/v1/login', { email, password, project: projectThatIsNotYours })
      .reply(404, errorResponse);

    await expect(run()).rejects.toStrictEqual(new APIv1Error(errorResponse));
    mock.done();
  });
});
