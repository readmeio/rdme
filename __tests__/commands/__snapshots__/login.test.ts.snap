// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rdme login > should bypass prompts and post to /login on the API if passing in every opt (no 2FA) 1`] = `
{
  "result": "Successfully logged in as user@example.com to the subdomain project.",
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should bypass prompts and post to /login on the API if passing in every opt 1`] = `
{
  "result": "Successfully logged in as user@example.com to the subdomain project.",
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should error if email is invalid 1`] = `
{
  "error": [Error: You must provide a valid email address.],
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should error if invalid credentials are given 1`] = `
{
  "error": [APIv1Error: Either your email address or password is incorrect

If you need help, email support@readme.io and mention log "fake-metrics-uuid".],
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should error if no project provided 1`] = `
{
  "error": [Error: No project subdomain provided. Please use \`--project\`.],
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should error if trying to access a project that is not yours 1`] = `
{
  "error": [APIv1Error: The project (unauthorized-project) can't be found.

If you need help, email support@readme.io],
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should make additional prompt for token if login requires 2FA 1`] = `
{
  "result": "Successfully logged in as user@example.com to the subdomain project.",
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should post to /login on the API 1`] = `
{
  "result": "Successfully logged in as user@example.com to the subdomain project.",
  "stderr": "",
  "stdout": "",
}
`;

exports[`rdme login > should post to /login on the API if passing in project via opt 1`] = `
{
  "result": "Successfully logged in as user@example.com to the subdomain project.",
  "stderr": "",
  "stdout": "",
}
`;
